test-plan/eventing/subscription-lifecycle-conformance.md

```diff
@@ -0,0 +1,257 @@
+- A [Channel resource that is referenced by the Subscription](subscription-lifecycle/channel.yaml)
+- A [Subscription resource that reference a Channel in a different namespace](control-plane/subscription-lifecycle/subscription-channel-namespace.yaml)
+- A [Subscription resource that has no Subscriber defined in the Spec](control-plane/subscription-lifecycle/subscription-no-subscriber.yaml)
+- A [Services resource that creates the DeadLetterSink and the Subscriber for the Subscription](control-plane/subscription-lifecycle/services.yaml)
+
+
+## [Pre] Creating a Subscription 
+
+```
+kubectl apply -f control-plane/subscription-lifecycle/subscription.yaml
+```
+
+## [Test] Inmutability
+
+Check for default annotations, this should return the name of the selected implementation: 
+
+```
+kubectl get subscription conformance-subscription -o jsonpath='{.spec.channel}'
+```
+
+Try to patch the annotation: `spec.channel` to see if the resource mutates: 
+
+```
+kubectl patch subscription conformance-subscription --type merge -p '{"spec":{"channel":{"apiVersion":"mutable"}}}'
+```
+
+You should get the following error: 
+```
+Error from server (BadRequest): admission webhook "validation.webhook.eventing.knative.dev" denied the request: validation failed: Immutable fields changed (-old +new): spec
+{v1.SubscriptionSpec}.Channel.APIVersion:
+	-: "messaging.knative.dev/v1"
+	+: "mutable"
+```
+
+Tested in eventing:
+- 
+
+### [Output]
+
+```
+{
+  "test": "control-plane/subscription-lifecycle/immutability"
+  "output": {
+	  "expectedError": "<EXPECTED_ERROR>"
+  }
+}
+```
+
+## [Test] Subscription Readiness 
+
+Check for condition type `Ready` with status `False` cause the some of the preconditions on the spec are not met: 
+
+```
+ kubectl get subscription conformance-subscription -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].status}"
+```
+
+Lets now check the reason why the Subscription is not ready yet:
+
+```
+ kubectl get subscription conformance-subscription -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].reason}"
+```
+
+Tested in eventing:
+- 
+
+### [Output]
+
+```
+{
+  "test": "control-plane/subscription-lifecycle/subscription-readiness"
+  "output": {
+    "expectedType": "Ready",
+    "expectedStatus": "Unknown",
+    "expectedReason: "ChannelReferenceFailed"
+  }
+}
+```
+
+## [Pre] Creating the Channel
+
+```
+kubectl apply -f control-plane/subscription-lifecycle/channel.yaml
+```
+
+## [Test] Subscription Readiness 2
+
+Check for condition type `Ready` with status `False` cause even with the Channel some of the preconditions on the spec are not met: 
+
+```
+ kubectl get subscription conformance-subscription -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].status}"
+```
+
+Lets now check the reason why the Subscription is not ready yet:
+
+```
+ kubectl get subscription conformance-subscription -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].reason}"
+```
```

Comment:
Conformance states that at least one condition MUST indicate the reason for the error.
I think that checking the at least one of the other conditions is set to false would make this conformant. Would there by some kubectl + JSON able to test that?

Reply:
Also, at this stage it should be checked that `status.physicalSubscription` is empty